core/src/proxy/auth.rs

```diff
@@ -42,206 +40,160 @@ impl Interceptor for AuthInterceptor {
+            client_pubkey: keypair.pubkey().as_ref().to_vec(),
+            signed_challenge,
+        })
+        .await
+        .map_err(|e| ProxyError::AuthenticationError(e.to_string()))?;
+
+    let inner = auth_tokens.into_inner();
+    let access_token = get_validated_token(inner.access_token)?;
+    let refresh_token = get_validated_token(inner.refresh_token)?;
+
+    Ok((access_token, refresh_token))
+}
 
-    /// Invokes the refresh_access_token gRPC method.
-    /// Returns a new access_token.
-    async fn refresh_access_token(
-        auth_service_client: &mut AuthServiceClient<Channel>,
-        refresh_token: Token,
-    ) -> crate::proxy::Result<Token> {
-        match auth_service_client
-            .refresh_access_token(RefreshAccessTokenRequest {
-                refresh_token: refresh_token.value,
-            })
-            .await
-        {
-            Ok(resp) => get_validated_token(resp.into_inner().access_token),
-            Err(e) => Err(ProxyError::GrpcError(e)),
-        }
-    }
+/// Tries to refresh the access token or run full-reauth if needed.
+/// This method writes to access_token if refresh is done.
+/// It overwrites the refresh token if full-reauth is run.
+pub async fn maybe_refresh_auth_tokens(
+    auth_service_client: &mut AuthServiceClient<Channel>,
+    auth_uri: String,
+    access_token: &Arc<Mutex<Token>>,
+    refresh_token: &mut Token,
+    cluster_info: &Arc<ClusterInfo>,
+    connection_timeout: &Duration,
+    auth_refresh_lookahead: u64,
+    num_full_refreshes: &mut u64,
+    num_refresh_access_token: &mut u64,
+) -> crate::proxy::Result<()> {
+    let access_token_expiry: u64 = access_token
+        .lock()
+        .unwrap()
+        .expires_at_utc
+        .as_ref()
+        .map(|ts| ts.seconds as u64)
+        .unwrap_or_default();
+    let refresh_token_expiry: u64 = refresh_token
+        .expires_at_utc
+        .as_ref()
+        .map(|ts| ts.seconds as u64)
+        .unwrap_or_default();
+
+    let now = Utc::now().timestamp() as u64;
+
+    let should_refresh_access =
+        access_token_expiry.checked_sub(now).unwrap_or_default() <= auth_refresh_lookahead;
+    let should_generate_new_tokens =
+        refresh_token_expiry.checked_sub(now).unwrap_or_default() <= auth_refresh_lookahead;
+
+    if should_generate_new_tokens {
+        let kp = cluster_info.keypair().clone();
+
+        let (new_access_token, new_refresh_token) = timeout(
+            *connection_timeout,
+            generate_auth_tokens(auth_service_client, kp.as_ref()),
+        )
+        .await
+        .map_err(|_| ProxyError::MethodTimeout("refresh_access_token".to_string()))?
+        .map_err(|e| ProxyError::MethodError(e.to_string()))?;
+
+        *access_token.lock().unwrap() = new_access_token.clone();
+        *refresh_token = new_refresh_token.clone();
+
+        *num_full_refreshes += 1;
+        datapoint_info!(
+            "auth_tokens_update_loop-tokens_generated",
```

Comment:
let's pass this metric name down from above and have it be prepended with relayer or block-engine depending on what stage is invoking

Reply:
in that case, do we need the url?